sphinx/domains/__init__.py

```diff
@@ -287,10 +289,12 @@ def run(self) -> List[Node]:
 
     # methods that should be overwritten
 
+    @abstractmethod
     def clear_doc(self, docname: str) -> None:
         """Remove traces of a document in the domain-specific inventories."""
         pass
 
+    @abstractmethod
     def merge_domaindata(self, docnames: List[str], otherdata: Dict) -> None:
```

Comment:
Note: This changes 3rd party domains must support parallel builds.

Reply:
This is a new constraint. So far, extension developers have been able to choose not to support parallel builds. But this enforce them to implement it. It must be a breaking change. I'll revert this.